dali/kernels/kernel_manager.h

```diff
@@ -0,0 +1,237 @@
+  ///          scratch memory. While the function returns non-const reference, please note
+  ///          that decreasing scratch sizes calculated by Setup will result in undefined
+  ///          behavior, including memory corruption or illegal access.
+  template <typename Kernel, typename... InArgs>
+  KernelRequirements &Setup(int instance_idx, KernelContext &context, InArgs &&...in_args) {
+    auto &inst = instances[instance_idx];
+    inst.requirements = inst.get<Kernel>().Setup(context, std::forward<InArgs>(in_args)...);
+    return inst.requirements;
+  }
+
+  /// @brief Calls Run on specified kernel instance using Scratchpad for given thread.
+  ///
+  /// @param thread_idx     - zero-based thread index
+  /// @param instance_idx   - kernel instance index; typically corresponds
+  ///                         to sample index (for per-sample kernels) or minibatch index
+  /// @param context        - context for the kernel
+  ///                         * should contain valid CUDA stream for GPU kernels;
+  ///                         * scratchpad pointer is overriden with a scratchpad
+  ///                           created for given thread_idx
+  /// @param out_in_args    - pack of arguments (outputs, inputs, arguments) used in Kernel::Run
+  template <typename Kernel, typename... OutInArgs>
+  void Run(int thread_idx, int instance_idx, KernelContext &context, OutInArgs &&...out_in_args) {
+    assert(static_cast<size_t>(thread_idx) < scratchpads.size());
+    auto &sa = GetScratchpadAllocator(thread_idx);
+    Run<Kernel>(sa, instance_idx, context, std::forward<OutInArgs>(out_in_args)...);
+  }
+
+  /// @brief Calls Run on specified kernel instance using Scratchpad for given thread.
+  ///
+  /// @param sa             - scratchpad allocator; memory will be reserved in it to satisfy
+  ///                         instance's requirements
+  /// @param instance_idx   - kernel instance index; typically corresponds
+  ///                         to sample index (for per-sample kernels) or minibatch index
+  /// @param context        - context for the kernel
+  ///                         * should contain valid CUDA stream for GPU kernels;
+  ///                         * scratchpad pointer is overriden with a scratchpad
+  ///                           created from `sa`
+  /// @param out_in_args    - pack of arguments (outputs, inputs, arguments) used in Kernel::Run
+  template <typename Kernel, typename... OutInArgs>
+  void Run(ScratchpadAllocator &sa,
+           int instance_idx,
+           KernelContext &context,
+           OutInArgs &&...out_in_args) {
+    assert(instance_idx >= 0 &&
+           static_cast<size_t>(instance_idx) < NumInstances() &&
+           "Kernel instance index (instance_idx) out of range");
+    auto &inst = instances[instance_idx];
+    ReserveScratchpad(sa, inst.requirements.scratch_sizes);
+    auto scratchpad = sa.GetScratchpad();
+    auto *old_scratchpad = context.scratchpad;
```

Comment:
Do you expect to have some nested calls?

Reply:
Nested - no, but I can imagine reusing the context. This one extra assignment and a temporary value won't make much difference in terms of performance and can save a lot of headaches should someone at some point decide to use the same context in raw Kernel API and with the Manager.